sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp

```diff
@@ -0,0 +1,365 @@
+      std::string EncryptionKeySha256 = std::string();
+    };
+
+    struct StageBlockOptions : public BodiedRequestOptions
+    {
+      std::string BlockId;
+      std::string ContentMd5 = std::string();
+      std::string ContentCrc64 = std::string();
+      std::string LeaseId = std::string();
+      std::string EncryptionKey = std::string();
+      std::string EncryptionKeySha256 = std::string();
+      std::string EncryptionScope = std::string();
+    };
+
+    static azure::core::http::Request StageBlockConstructRequest(
+        /*const*/ std::string& url,
+        /*const*/ StageBlockOptions& options)
+    {
+      azure::core::http::Request request(azure::core::http::HttpMethod::Put, url);
+      request.addQueryParameter(k_BLOCK_ID, options.BlockId);
+      request.addQueryParameter(k_QUERY_COMP, k_QUERY_BLOCK);
+      request.setBodyBuffer(options.BodyBuffer);
+      request.setBodyStream(options.BodyStream);
+      ApplyBasicHeaders(options, request);
+      return request;
+    }
+
+    // TODO: should return azure::core::http::Response<BlockInfo> instead
+    static BlockInfo StageBlockParseResponse(/*const*/ azure::core::http::Response& response)
+    {
+      BlockInfo info;
+      info.ContentMd5 = GetHeaderValue(response.getHeaders(), k_HEADER_CONTENT_MD5);
+      info.ContentCrc64 = GetHeaderValue(response.getHeaders(), k_HEADER_X_MS_CONTENT_CRC64);
+      ParseBasicResponseHeaders(response.getHeaders(), info);
+      return info;
+    }
+
+    // TODO: should return azure::core::http::Response<BlockInfo> instead
+    static BlockInfo StageBlock(
+        // TODO: Context and Pipeline should go here
+        /*const*/ std::string& url,
+        /*const*/ StageBlockOptions& options)
+    {
+      return StageBlockParseResponse(
+          azure::core::http::Client::send(StageBlockConstructRequest(url, options)));
+    }
+
+    struct StageBlockListOptions : public RequestOptions
+    {
+      std::vector<std::pair<BlockType, std::string>> BlockList
+          = std::vector<std::pair<BlockType, std::string>>();
+      std::string ContentMd5 = std::string();
+      std::string ContentCrc64 = std::string();
+      std::string LeaseId = std::string();
+      std::string EncryptionKey = std::string();
+      std::string EncryptionKeySha256 = std::string();
+      std::string EncryptionScope = std::string();
+    };
+
+    static azure::core::http::Request StageBlockListConstructRequest(
+        /*const*/ std::string& url,
+        /*const*/ StageBlockListOptions& options)
+    {
+      azure::core::http::Request request(azure::core::http::HttpMethod::Put, url);
+      request.addQueryParameter(k_QUERY_COMP, k_QUERY_BLOCK_LIST);
+      ApplyBasicHeaders(options, request);
+      //TODO: avoid the data copy here.
+      block_list_writer writer;
+      std::string body = writer.write(options.BlockList);
+      uint8_t* buffer = new uint8_t[body.size() + 1];
+      memcpy_s(buffer, body.size() + 1, body.c_str(), body.size() + 1);
```

Comment:
Is `memcpy_s` in C++ standard?

Reply:
If this is product code, std::copy will be used as we offline discussed. However, this is unlikely to be product code since the memory management here is not ideal.

Reply:
@JinmingHu-MSFT  No it is not.